core/shared/src/main/scala/zio/FiberRefs.scala

```diff
@@ -26,50 +27,78 @@ import scala.annotation.tailrec
  * example between an asynchronous producer and consumer.
  */
 final class FiberRefs private (
-  private[zio] val fiberRefLocals: Map[FiberRef[_], (::[(FiberId.Runtime, Any, Int)], Int)]
+  private[zio] val fiberRefLocals: Map[FiberRef[_], FiberRefs.Value]
 ) { self =>
+  import zio.FiberRefs.{StackEntry, Value}
 
   /**
    * Returns a new fiber refs with the specified ref deleted from it.
    */
-  def delete(fiberRef: FiberRef[_]): FiberRefs =
-    FiberRefs(fiberRefLocals - fiberRef)
+  def delete(fiberRef: FiberRef[_]): FiberRefs = {
+    val newMap = fiberRefLocals - fiberRef
+    if (newMap eq fiberRefLocals) self
+    else FiberRefs(newMap)
+  }
 
   /**
    * Returns a set of each `FiberRef` in this collection.
    */
   def fiberRefs: Set[FiberRef[_]] =
     fiberRefLocals.keySet
 
+  /**
+   * Boolean flag which indicates whether the FiberRefs map contains an entry
+   * that will cause the map to be changed when [[forkAs]] is called.
+   *
+   * This way we can avoid calling `fiberRefLocals.transform` on every
+   * invocation of [[forkAs]] when we already know that the map will not be
+   * transformed
+   */
+  @volatile private var needsTransformWhenForked: Boolean = true
+
   /**
    * Forks this collection of fiber refs as the specified child fiber id. This
    * will potentially modify the value of the fiber refs, as determined by the
    * individual fiber refs that make up the collection.
    */
-  def forkAs(childId: FiberId.Runtime): FiberRefs = {
-    val childFiberRefLocals: Map[FiberRef[_], (::[(FiberId.Runtime, Any, Int)], Int)] =
-      fiberRefLocals.transform { case (fiberRef, (stack, depth)) =>
-        val oldValue = stack.head._2.asInstanceOf[fiberRef.Value]
-        val newValue = fiberRef.patch(fiberRef.fork)(oldValue)
-        if (oldValue == newValue) (stack, depth) else (::((childId, newValue, 0), stack), depth + 1)
+  def forkAs(childId: FiberId.Runtime): FiberRefs =
+    if (needsTransformWhenForked) {
+      val childFiberRefLocals =
+        fiberRefLocals.transform { case (fiberRef, entry @ Value(stack, depth)) =>
+          val oldValue = stack.head.value.asInstanceOf[fiberRef.Value]
+          val newValue = fiberRef.patch(fiberRef.fork)(oldValue)
+          if (oldValue == newValue) entry
+          else Value(::(StackEntry(childId, newValue, 0), stack), depth + 1)
+        }
+      if (fiberRefLocals eq childFiberRefLocals) {
```

Comment:
I think you need `.equals` here and not `.eq`, as I am sure Scala collections will build a brand new map every time with `transform`. Note that comparing maps for equality involves lots of allocations and virtual invocations.

Reply:
I checked the implementation for `HashMap`, `Map1`, `Map2`, etc and they all seem to return the same object when `transform` is called if the map is unchanged. I also checked it in the REPL and it seems to be the case. e.g., for `HashMap`:

<img width="597" alt="image" src="https://github.com/zio/zio/assets/67301607/83d7383b-8adc-4275-ad27-0e4c9396faf3">

Although as @guizmaii [pointed out](https://github.com/zio/zio/pull/8745#discussion_r1567002879), since these optimizations rely very heavily on the underlying implementation remaining the same, I'll add a simple test for making sure this behaviour doesn't change in the future.

Also I need to check if my above statement is also correct for Scala 2.12 🤔

Reply:
Turns out `transform` returns a new instance of the Map in Scala 2.12. To make the optimizations compatible without using `equals`, I added a local var that is set to `true` when a transformation takes place inside `transform`.

Reply:
FYI, in zio-kafka, we've dropped Scala 2.12 support for sometimes already (https://github.com/zio/zio-kafka/releases/tag/v2.7.0)
I know it's a bit radical, but we could consider making the same decision for ZIO

Reply:
Yeah I think besides Spark, there are also issues with SBT plugins since those are still written in Scala 2.12. Caliban's SBT plugin depends on ZIO, so if we were to drop Scala 2.12 support we wouldn't be able to update ZIO until we can write SBT plugins in Scala 2.13 / 3